File: DSA/DSA06036 BỘ BA SỐ BẰNG K.cpp
#include <bits/stdc++.h>
using namespace std;

void submit()
{
    int n, k;
    cin >> n >> k;
    vector<int> A(n);
    for (auto &&i : A) cin >> i;
    
    sort(A.begin(), A.end());
    bool check = false;

    for (int i = 0; i < n-1; i++)
    {
        int l = i+1, r = n-1;

        while (l < r)
        {
            if (A[l] + A[r] < k - A[i])
            {
                l++;
            }
            else if (A[l] + A[r] > k - A[i])
            {
                r--;
            }
            else
            {
                check = true;
                break;
            }
        }
    }
    
    check ? cout << "YES" : cout << "NO";
    
    


    cout << endl;
}


int main()
{
    ios::sync_with_stdio(false);                                    // Speed down input/output
    cin.tie(nullptr);
    cout.tie(nullptr);
    
    int so_bo_test;
    cin >> so_bo_test;
    cin.ignore();
    
    while (so_bo_test--)
    {
        submit();
    }

    return 0;
}

File: DSA_python/DSA06036 BỘ BA SỐ BẰNG K.py
from bisect import bisect_right

def submit():
    n, k = map(int, input().split())
    A = list(map(int, input().split()))

    A.sort()
    check = False

    for i in range(n-2):
        temp = i+1
        j = bisect_right(A, k-A[i]-A[temp])-1
        p = k-A[i]
        while j > temp:
            if A[temp]+A[j] < p:
                temp+=1
            elif A[temp]+A[j] > p:
                j-=1
            else:
                check = True
                break
        if check == True: break
    
    print("YES" if check == True else "NO")

test = int(input())
for i in range(test):
    submit()
